BridgeAppSDK/SBAGenericStepDataSource.swift

```diff
@@ -0,0 +1,614 @@
+            }
+            
+            parentResult.results = parentResult.results! + qResults
+        }
+        
+        return parentResult
+    }
+    
+    fileprivate func formItemsWithAnswerFormat() -> Array<ORKFormItem>? {
+        
+        if let formItems = formItems() {
+            var items: Array<ORKFormItem> = Array()
+            formItems.forEach({
+                if let _ = $0.answerFormat {
+                    items.append($0)
+                }
+            })
+            return items
+        }
+        return nil
+    }
+    
+    fileprivate func formItems() -> Array<ORKFormItem>? {
+        switch self.step {
+        case is ORKFormStep:
+            return (self.step as! ORKFormStep).formItems
+        case is ORKQuestionStep:
+            return (self.step as! ORKQuestionStep).formItems
+        default:
+            return nil
+        }
+    }
+    
+    fileprivate func populate() {
+        
+        if let items = formItems(), items.count > 0 {
+            
+            let singleSelectionTypes: Array<Int> = [ORKQuestionType.boolean.rawValue,
+                                                    ORKQuestionType.singleChoice.rawValue,
+                                                    ORKQuestionType.multipleChoice.rawValue,
+                                                    ORKQuestionType.location.rawValue]
+            
+            var newSection: SBAGenericStepTableSection? = nil
+            
+            for item in items {
+                
+                // In case no section available, create new one.
+                if newSection == nil {
+                    newSection = SBAGenericStepTableSection(sectionIndex: sections.count)
+                }
+                
+                if let answerFormat = item.answerFormat?.implied() {
+                    let multiCellChoices = singleSelectionTypes.contains(answerFormat.questionType.rawValue) && type(of: answerFormat) != ORKValuePickerAnswerFormat.self
+                    let multiLineTextEntry = answerFormat.questionType == .text
+                    let scale = answerFormat.questionType == .scale
+                    
+                    
+                    if multiCellChoices || multiLineTextEntry || scale {
+                        
+                        newSection!.title = item.text
+                        newSection!.add(formItem: item)
+                        sections.append(newSection!)
+                        
+                        // following item should start a new section
+                        newSection = nil
+                    }
+                    else {
+                        
+                        newSection!.add(formItem: item)
+                        sections.append(newSection!)
+                    }
+                }
+                else {
+                    
+                    newSection!.title = item.text
+                    sections.append(newSection!)
```

Comment:
Yeah, this section was one of the first things I worked on and I copied this code directly from RK ((void)buildSections in ORKFormStepVC), thinking it was safest to do exactly what RK was doing. I agree that it's not well written. I'll redo it.